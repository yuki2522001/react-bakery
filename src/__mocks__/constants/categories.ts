import { CategoryProps } from "@common-types/category";

const CATEGORY_MOCKING_LIST: CategoryProps[] = [
  {
    id: "1",
    name: "Cake"
  },
  {
    id: "2",
    name: "Breads"
  }
]

export { CATEGORY_MOCKING_LIST };
