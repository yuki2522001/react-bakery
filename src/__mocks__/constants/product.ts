import { Product } from "@common-types/product";
import url from "../../assets/images/products/black-forest-cake.jpg";

const PRODUCT_MOCKING: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    url
  ]
}

const PRODUCT_MOCKING_LIST: Product[] = [
  {
    id: "1",
    categoryId: "1651924356334",
    name: "Cheese cake",
    price: 20,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  },
  {
    id: "2",
    categoryId: "1651924356334",
    name: "Amaretti Cookie",
    price: 10,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  }
]

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
