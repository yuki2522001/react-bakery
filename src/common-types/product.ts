import { CategoryProps } from "./category";

export interface Product {
  id: string;
  name: string;
  price: number;
  images: string[];
  quantity: number;
  categoryId: string;
  description: string;
  product?: {};
}

export interface ProductContext {
  products: Product[];
  dispatch: Function,
  searchValue: string;
  setSearchValue: Function;
  categories: CategoryProps[];
  setCategories: Function;
}