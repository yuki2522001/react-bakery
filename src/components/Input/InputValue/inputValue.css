.input__value {
  padding: 7px 10px;
  font-size: 13px;
  border: 1px solid var(--border-btn);
  border-radius: 3px;
}

.input__number {
  padding: 7px 0;
  width: 150px;
  border: 1px solid var(--border-btn);
  margin-right: 10px;
  border-radius: 3px;
}

.input__text:hover,
.input__number:hover,
.input__value:hover {
  border-color: var(--text-primary);
}