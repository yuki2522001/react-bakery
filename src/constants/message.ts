const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

const ERROR_MSG = {
  REQUIRED: "Field is required",
  NUMBER: "Enter a non-negative number"
}

export { SUCCESS_MSG, ERROR_MSG };
