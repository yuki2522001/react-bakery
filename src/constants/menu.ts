import { MenuType } from "../common-types/menu";

const MENU_LIST: MenuType[] = [
  {
    key: "about",
    label: "About",
  },
  {
    key: "shop",
    label: "Shop",
  },
  {
    key: "blog",
    label: "Blog",
  },
  {
    key: "faq",
    label: "FAQ",
  },
  {
    key: "contact",
    label: "Contact"
  }
];

export default MENU_LIST;
