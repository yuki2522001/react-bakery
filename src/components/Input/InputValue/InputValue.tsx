import React, { ChangeEventHandler, memo } from "react";
import "./inputValue.css"

interface InputProps {
  id?: string;
  className?: string;
  value?: string | number;
  type?: string;
  name?: string;
  min?: number;
  multiple?: boolean;
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

const InputValue: React.FC<InputProps> = memo(({
    id,
    className,
    value,
    type,
    min,
    name,
    multiple,
    onChange,
  }) => {
    return (
      <input
        id={id}
        className={className}
        value={value}
        type={type}
        min={min}
        name={name}
        multiple={multiple}
        onChange={onChange}
      />
    );
  }
);

export default InputValue;
