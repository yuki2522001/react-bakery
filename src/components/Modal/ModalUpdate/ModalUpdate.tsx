import useSWR from "swr";
import React, { ChangeEvent, useCallback, useContext, useState } from "react";
import toast from "react-hot-toast";
import getBase64 from "@helpers/getBase64";
import InputValue from "@components/Input/InputValue/InputValue";
import Button from "@components/common/Button/Button/Button";
import { FieldName, FormProps } from "@common-types/form";
import { Product } from "@common-types/product";
import { RULES } from "@constants/rules";
import { validate } from "@helpers/validate";
import { CategoryProps } from "@common-types/category";
import { SUCCESS_MSG } from "@constants/message";
import { getData, update } from "@helpers/apiHandle";
import { setFieldsValue } from "@helpers/fieldHandle";
import { CATEGORIES_URL, PRODUCTS_URL } from "@constants/url";
import { DataContext } from "@context/DataContext";
import { Action } from "@common-types/data";
import "../modal.css";

interface ModalUpdateProps {
  product: Product;
  hideModalUpdate: () => void;
  deleteImage: () => void;
  updateProductDetail: (product: Product) => void;
}

const ModalUpdate: React.FC<ModalUpdateProps> = ({
  product,
  hideModalUpdate,
  updateProductDetail,
}) => {
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<CategoryProps[]>);
  const {dispatch} = useContext(DataContext);
  // create state to handle select file image
  const [selectedFile, setSelectedFile] = useState<string[]>([]);
  // create state to update product
  const [productEdit, setProductEdit] = useState<Product>(product);
  const updateProduct = async (id: string, productData: Product) => {
    const productEdit: Product = {
      id: productData.id,
      categoryId: productData.categoryId,
      name: productData.name,
      price: +productData.price,
      quantity: +productData.quantity,
      description: productData.description,
      images: productData.images,
    };
    try {
      const response = await update(`${PRODUCTS_URL}/${id}`, productEdit);
      if(response) {
        dispatch({
          action: Action.UpdateProductSuccess,
          payload: {...response.data}
        });
        updateProductDetail(response.data);
      }
      toast.success(SUCCESS_MSG.MESSAGE_UPDATE_PRODUCT);
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  const [formValues, setFormValues] = useState<FormProps>({
    categoryId: {
      value: productEdit?.categoryId,
      rules: [RULES.REQUIRED],
      error: "",
    },
    name: {
      value: productEdit?.name,
      rules: [RULES.REQUIRED],
      error: "",
    },
    price: {
      value: productEdit?.price.toString(),
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    quantity: {
      value: productEdit?.quantity.toString(),
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    description: {
      value: productEdit?.description,
      rules: [RULES.REQUIRED],
      error: "",
    },
    images: {
      value: productEdit?.images.toString(),
      rules: [],
      error: "",
    }
  });

  // handle update product
  const handleUpdateProduct = (id: string, productEdit: Product) => {
    for (let i = 0; i < selectedFile.length; i++) {
      productEdit.images.push(selectedFile[i]);
    }

    // validate form
    setFormValues(validate(formValues));
    const temp = (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        if (formValues[fieldName].error) {
          return false;
        }
      }
    );

    if (!temp.includes(false)) {
      setFormValues({...formValues});
      updateProduct(id, productEdit);
      hideModalUpdate();
    }
  };

  // handle change value
  const handleChange = useCallback((event: { target: { value: string; name: string } }) => {
    const value = event.target.value;
    const fieldName = event.target.name as FieldName;
    setProductEdit({ ...productEdit, [fieldName]: value });
    setFormValues(setFieldsValue(formValues, value, fieldName));
  }, [productEdit]);

  // handle change image
  const imageChange = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const files = event.target.files;
    const newFiles = [];
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        newFiles.push(imageSrc)
      }
    }
    setSelectedFile([...selectedFile, ...newFiles]);
  }, [selectedFile]);

  // handle delete image
  const handleDeleteImage = useCallback((event: { target: EventTarget }) => {
    const target = event.target as Element;
    const indexOfArr = productEdit.images.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    const indexOf = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOf, 1);
    productEdit.images.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  },[selectedFile]);

  return (
    <div data-testid="modal-update" className="modal-update">
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Edit Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name: </label>
              <InputValue
                type="text"
                name="name"
                value={productEdit?.name}
                onChange={handleChange}
              />
              <small className="form__error">
                {formValues.name.error}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Description: </label>
              <textarea
                data-testid="change-value"
                className="input__text"
                cols={30}
                rows={5}
                value={productEdit?.description}
                name="description"
                onChange={handleChange}
              ></textarea>
              <small className="form__error">
                {formValues.description.error}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories: </label>
              <select
                className="form__select"
                name="categoryId"
                onChange={handleChange}
              >
                {data?.map(({ id, name }: CategoryProps, index: number) => (
                  <option
                    key={index}
                    value={id}
                  >
                    {name}
                  </option>
                ))}
              </select>
              <small className="form__error">
                {formValues.categoryId.error}
              </small>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price: </label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  value={productEdit?.price}
                  name="price"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.price.error}
                </small>
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity: </label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  value={productEdit?.quantity}
                  name="quantity"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.quantity.error}
                </small>
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {productEdit?.images?.map((img: string, index: number) => (
                  <img
                    data-testid="delete-image"
                    key={index}
                    className="form__img"
                    data-id={img}
                    src={img}
                    onClick={handleDeleteImage}
                  />
                ))}
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number) => (
                    <img
                      data-testid="after-change-file"
                      key={key}
                      className="form__img"
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <Button
              type="warning"
              onClick={hideModalUpdate}
              text="Cancel"
            />
            <Button
              text="Submit"
              type="success"
              onClick={() => handleUpdateProduct(product.id, productEdit)}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalUpdate;
