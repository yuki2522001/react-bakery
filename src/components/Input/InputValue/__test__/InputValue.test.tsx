import ModalUpdate from "@components/Modal/ModalUpdate/ModalUpdate";
import { PRODUCT_MOCKING } from "@__mocks__/constants/product";
import { fireEvent, render } from "@testing-library/react";
import InputValue from "../InputValue";

describe("Input value component", () => {
  const onChangeValue = jest.fn();
  const setup = () => {
    const utils = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    const input = utils.getByTestId("change-value") as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };
  test("Should onChange value when input value", async () => {
    render (
      <InputValue onChange={onChangeValue} />
    );
    const { input } = setup();
    fireEvent.change(input, { target: { value: "Cheese pocket" } });
    expect(input.value).toBe("Cheese pocket");
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<InputValue />);
    expect(asFragment()).toMatchSnapshot();
  });
});