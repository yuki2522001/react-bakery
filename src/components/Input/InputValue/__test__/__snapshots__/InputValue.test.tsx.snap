// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input value component matches snapshot 1`] = `
<DocumentFragment>
  <input
    value=""
  />
</DocumentFragment>
`;
